### PR TITLE
Core: Implement exists api

### PR DESCRIPTION
**Pull request by yanghua**
### Summary :memo:
_Write an overview about it._

### Details
_Describe more what you did on changes._
1. (...)
2. (...)

### Bugfixes :bug: (delete if dind't have any)
-

### Checks
- [ ] Closed #798
- [ ] Tested Changes
- [ ] Stakeholder Approval

### CODE REVIEW

File: tosfs/core.py [hunk 1]
```diff
@@ -49,6 +49,9 @@ def setup_logging() -> None:
     "The tosfs's log level is set to be %s", logging.getLevelName(logger.level)
 )
 
+# constants
+SERVER_RESPONSE_CODE_NOT_FOUND = 404
```
Comment (openinx) [inline tosfs/core.py:53]: Could we define those constant values into a separate `const.py` file ?

File: tosfs/core.py [hunk 2]
```diff
@@ -413,6 +416,200 @@ def _try_dir_info(self, bucket: str, key: str, path: str, fullpath: str) -> dict
         except Exception as e:
             raise TosfsError(f"Tosfs failed with unknown error: {e}") from e
 
+    def exists(self, path: str, **kwargs: Union[str, bool, float, None]) -> bool:
```
Comment (openinx) [inline tosfs/core.py:419]: What does this `**kwargs: Union[str, bool, float, None]` mean ?  What's the value that people should pass ?

Comment (openinx) [reply tosfs/core.py:419]: Souds like it's not quite clear for me to construct the argumenst when i call this method....

File: tosfs/core.py [hunk 3]
```diff
@@ -413,6 +416,200 @@ def _try_dir_info(self, bucket: str, key: str, path: str, fullpath: str) -> dict
...
+        """Check if a path exists in the TOS file system.
+
+        Parameters
+        ----------
+        path : str
+            The path to check for existence.
+        **kwargs : dict, optional
+            Additional arguments.
```
Comment (openinx) [inline tosfs/core.py:427]: The pydoc says it will be a dict, while the declared arg is a `Union` ?  Is this correct ?

File: tosfs/core.py [hunk 4]
```diff
@@ -413,6 +416,200 @@ def _try_dir_info(self, bucket: str, key: str, path: str, fullpath: str) -> dict
...
+        --------
+        >>> fs = TosFileSystem()
+        >>> fs.exists("tos://bucket/to/file")
+        True
+        >>> fs.exists("tos://mybucket/nonexistentfile")
+        False
+
+        """
+        if path in ["", "/"]:
+            # the root always exists, even if anon
```
Comment (openinx) [inline tosfs/core.py:453]: anon ? What's the meaning ?

File: tosfs/core.py [hunk 5]
```diff
@@ -413,6 +416,200 @@ def _try_dir_info(self, bucket: str, key: str, path: str, fullpath: str) -> dict
...
+        True
+        >>> fs._exists_bucket("nonexistentbucket")
+        False
+
+        """
+        if self.dircache.get(bucket, False):
+            return True
+        else:
+            try:
+                if self._ls_from_cache(bucket):
```
Comment (openinx) [inline tosfs/core.py:507]: ls_from_cache ? which means we will get a false positive result when accessing the the cache ?

Comment (openinx) [reply tosfs/core.py:507]: I mean if we've already removed the object key from backend object storage, while the cache is still holding the cached item, then we will regard the object key is still exising, which is incorrect ?

**Review by openinx — CHANGES_REQUESTED** [review]

File: tosfs/core.py (hunk 4 above)
Comment (yanghua) [reply tosfs/core.py:453]: `anon` means `anonymous `, aligned with `s3fs`.

File: tosfs/core.py (hunk 5 above)
Comment (yanghua) [reply tosfs/core.py:507]: Yes, there may be multi-source consistency issues here. The current cache can only guarantee that the cached data is invalidated when it performs the deletion operation by itself. To ensure accuracy, one way is not to adopt the caching mechanism or to pass the refresh parameter everywhere. WDYT?

Comment (yanghua) [reply tosfs/core.py:506]: After some thinking, I feel that the disadvantages of introducing the list cache outweigh the advantages for us. The main considerations are as follows: 

* We cannot assume that users read and write the TOS only through the single entry of TOSFS. The cost of maintaining cache consistency and the cost of misjudgment may be more serious compared to performance. 
* If we assume that the main usage scenario of the target users is programming, then FS-level APIs such as those dependent on caching like ls, exists, info... will not be used so frequently. 

After seeing that fsspec has built-in basic support for `dircache` and s3fs followed suit, I also introduced caching by default. But now I feel that not introducing caching seems to be a better option.  WDYT? @openinx @xianyinxin

Comment (yanghua) [reply tosfs/core.py:506]: @openinx After discussing with @xianyinxin , we agreed with removing cache logic from tosfs, the biggest concern is for some dangerous operations(e.g. `rm`), if misjudgments are caused by caching, the cost is very high.

**Review by xianyinxin — APPROVED** [review]

**Review by openinx — APPROVED** [review]

**Pull request merged**